packages/circuits/circom/poseidon-cipher.circom

```diff
@@ -0,0 +1,219 @@
+    var decryptedLength = length;
+    while (decryptedLength % 3 != 0) {
+        decryptedLength += 1;
+    }
+
+    // the ciphertext to decrypt
+    signal input ciphertext[decryptedLength + 1];
+    // nonce passed in while encrypting
+    signal input nonce;
+    // the decryption key 
+    signal input key[2];
+
+    signal output decrypted[decryptedLength];
+    signal output decryptedLast;
+
+    var two128 = 2 ** 128;
+
+    // nonce must be < 2^128
+    component lt = LessThan(252);
+    lt.in[0] <== nonce;
+    lt.in[1] <== two128;
+    lt.out === 1;
+
+    // calculate the number of iterations 
+    // needed for the decryption
+    // process
+    // \ is integer division
+    var n = (decryptedLength + 1) \ 3;
+
+    // we store our strategies in an array
+    component strategies[n+1];
+    // iterate poseidon on the initial state
+    strategies[0] = PoseidonPerm(4);
+
+    // we need to set the initial state to 
+    // [0, key[0], key[1], nonce + (length * 2^128)]
+    // so we create one extra component 
+    // and run a permutation so we can set the state
+    strategies[0].inputs[0] <== 0;
+    strategies[0].inputs[1] <== key[0];
+    strategies[0].inputs[2] <== key[1];
+    strategies[0].inputs[3] <== nonce + (length * two128);
+    
+    // loop for n iterations
+    for (var i = 0; i < n; i++) {
+        // release three elements of the message
+        for (var j = 0; j < 3; j++) {
+            decrypted[i * 3 + j] <== ciphertext[i * 3 + j] - strategies[i].out[j + 1];
+        }
+
+        // create a new PoseidonPerm component
+        // at the next index 
+        // because we already passed all values to the 
+        // first component
+        strategies[i + 1] = PoseidonPerm(4);
+        strategies[i + 1].inputs[0] <== strategies[i].out[0];
+
+        // set the inputs from the ciphertext 
+        for (var j = 0; j < 3; j++) {
+            strategies[i + 1].inputs[j + 1] <== ciphertext[i * 3 + j];
+        }
+    }
+    
+    // pass in the last
+    decryptedLast <== strategies[n].out[1];
+}
+
+// Perform poseidon permutation on a nInputs input
+// The difference with circomlib Poseidon implementation
+// is that this one returns all of the intermediate values
+template PoseidonPerm(nInputs) {
+    signal input inputs[nInputs];
+    signal output out[nInputs];
+
+    // Using recommended parameters from whitepaper https://eprint.iacr.org/2019/458.pdf (table 2, table 8)
+    // Generated by https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/calc_round_numbers.py
+    // And rounded up to nearest integer that divides by t
+    var N_ROUNDS_P[16] = [56, 57, 56, 60, 60, 63, 64, 63, 60, 66, 60, 65, 70, 60, 64, 68];
+    var nRoundsF = 8;
+    var nRoundsP = N_ROUNDS_P[nInputs - 2];
+    var C[nInputs*(nRoundsF + nRoundsP)] = POSEIDON_C_OLD(nInputs);
+    var M[nInputs][nInputs] = POSEIDON_M_OLD(nInputs);
+
+    component ark[nRoundsF + nRoundsP];
```

Comment:
Just a note that this part of the code can be improved for readability in the future.

Reply:
101%!!